src/overlay/OverlayManagerImpl.cpp

```diff
@@ -1115,4 +1123,165 @@ OverlayManagerImpl::updateFloodRecord(StellarMessage const& oldMsg,
+                    }
+                    else if (numDemanded >= MAX_RETRY_COUNT)
+                    {
+                        shouldDemand = false;
+                    }
+                    else
+                    {
+                        // Check if it's been a while since our last demand
+                        std::chrono::milliseconds sinceLastDemand =
+                            std::chrono::duration_cast<
+                                std::chrono::milliseconds>(
+                                now - mDemandHistory[code].second);
+                        if (sinceLastDemand >=
+                            numDemanded * cfg.FLOOD_DEMAND_BACKOFF_DELAY_MS)
+                        {
+                            mOverlayMetrics.mDemandTimeoutMeter.Mark();
+                            TracyPlot("overlay.demand.timeout",
+                                      static_cast<int64_t>(
+                                          mOverlayMetrics.mDemandTimeoutMeter
+                                              .count()));
+                            shouldDemand = true;
+                        }
+                    }
+
+                    if (shouldDemand)
+                    {
+                        demandMap[p.first].push_back(code);
+                        mDemandHistory[code].first++;
+                        mDemandHistory[code].second = now;
+                    }
+                    else
+                    {
+                        // We've already requested this code in this round of
+                        // demand.
+                        nextRetryMap[p.first].push(code);
+                    }
+                }
+            }
+            if (demandMap[p.first].size() < cfg.FLOOD_MAX_DEMAND_SIZE &&
+                (p.second->txHashCount() + mRetryMap[p.first].size() > 0))
+            {
+                nextPeers.push_back(p);
+            }
+        }
+        remPeers = nextPeers;
+        if (remPeers.empty())
+        {
+            break;
+        }
+    }
+    for (auto& p : peers)
+    {
+        while (!mRetryMap[p.first].empty())
+        {
+            nextRetryMap[p.first].push(mRetryMap[p.first].front());
+            mRetryMap[p.first].pop();
+        }
+    }
+    mRetryMap = nextRetryMap;
+    for (auto& p : peers)
+    {
+        p.second->sendTxDemand(demandMap[p.first]);
+    }
+
+    // Remove old codes from the demand history.
+    std::vector<Hash> oldCodes;
+    for (auto const& it : mDemandHistory)
+    {
+        std::chrono::milliseconds sinceLastDemand =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                now - it.second.second);
+        if (sinceLastDemand >= MAX_DEMAND_HISTORY_RETENTION_DURATION)
+        {
+            TracyPlot("pending-dmeand-removed-too-old",
+                      static_cast<int64_t>(1));
+            oldCodes.push_back(it.first);
+        }
+    }
+    std::for_each(oldCodes.begin(), oldCodes.end(),
```

Comment:
there's no need to remove codes separately like this, `erase` returns the iterator to the next element in the map, so you can do something like: https://github.com/stellar/stellar-core/blob/fb86014666cb30a6478aa0b3cf5071b4994001e9/src/work/BatchWork.cpp#L36

Reply:
Thanks, this was a great idea. I used the same idea for the main loop in `demand()` to make it a bit more readable.